src/lib/optimizer/strategy/abstract_rule.hpp

```diff
@@ -2,34 +2,54 @@
 
 #include <memory>
 #include <string>
+#include <unordered_set>
+#include <vector>
 
 namespace opossum {
 
-class AbstractLQPNode;
-class AbstractCostEstimator;
 class AbstractCardinalityEstimator;
+class AbstractCostEstimator;
+class AbstractLQPNode;
+class LQPSubqueryExpression;
 
 class AbstractRule {
  public:
   virtual ~AbstractRule() = default;
 
   /**
    * This function applies the concrete Optimizer Rule to an LQP.
-   * apply_to() is intended to be called recursively by the concrete rule.
-   * The optimizer will pass the immutable LogicalPlanRootNode to this function.
+   * The optimizer will pass the immutable LogicalPlanRootNode (@param lqp_root) to this function.
    */
-  virtual void apply_to(const std::shared_ptr<AbstractLQPNode>& root) const = 0;
+  virtual void apply(const std::shared_ptr<AbstractLQPNode>& lqp_root) const;
 
   std::shared_ptr<AbstractCostEstimator> cost_estimator;
 
  protected:
   /**
-   * Apply this rule to @param node's inputs and all subqueries in its expressions
+   * _apply_to() is intended to be called recursively by the concrete rule.
+   */
+  virtual void _apply_to(const std::shared_ptr<AbstractLQPNode>& lqp_root) const = 0;
+
+  /**
+   * Calls _apply_to() for each input of @param node.
    *
    * IMPORTANT: Takes a copy of the node ptr because applying this rule to inputs of this node might remove this node
    * from the tree, which might result in this node being deleted if we don't take a copy of the shared_ptr here.
    */
   void _apply_to_inputs(std::shared_ptr<AbstractLQPNode> node) const;  // NOLINT
 };
 
+// All SubqueryExpressions referencing the same LQP
+using SubqueryExpressionsByLQP =
+    std::vector<std::pair<std::shared_ptr<AbstractLQPNode>, std::vector<std::shared_ptr<LQPSubqueryExpression>>>>;
+
+/**
+ * Traverses @param node with all its (nested) subquery expressions to identify unique LQPs and the (multiple)
+ * SubqueryExpressions referencing each of these unique LQPs.
+ */
+void collect_subquery_expressions_by_lqp(SubqueryExpressionsByLQP& subquery_expressions_by_lqp,
```

Comment:
If this is a free-floating function that is needed by multiple cpp files, I'd move it to lqp_utils. Also, the interface is weird, which is not your fault. I'd expect it to return a result. Maybe rename the current implementation to `recursively_collect_subquery_expressions_by_lqp`, move it into an anonymous namespace in `lqp_utils.cpp` and make `collect_subquery_expressions_by_lqp` simply call it with an empty `SubqueryExpressionsByLQP`?